ocfweb/docs/docs/staff/backend/kubernetes.md

```diff
@@ -0,0 +1,190 @@
+bootstrapping and bolt OCF specific configurations on top of it.
+`puppetlabs-kubernetes` performs two crucial tasks:
+
+1. Installs `etcd`, `kubelet`, `kube-proxy`, and `kube-dns`, initializes the
+   cluster, and applies a networking backend.
+2. Generates the required [PKI for Kubernetes and etcd][kubernetes-pki].
+
+Do note that `puppetlabs-kubernetes` is still very much a work in progress. If
+you notice an issue in the module you are encouraged to write a patch and send
+it upstream.
+
+### Kubernetes PKI
+
+All the private keys and certs for the PKI are in the puppet private share, in
+`/opt/puppet/shares/private/kubernetes`. We won't go into detail of everything
+contained there, but much of Kubernetes and `etcd` communication is
+authenticated using client certificates. All the necessary items for workers
+are included in `os/Debian.yaml`, although adding a new master to the cluster
+requires re-running [kubetool][puppetlabs-kubetool] to generate new `etcd
+server` and `etcd peer` certs.
+
+### OCF Kubernetes Configuration
+
+Currently, the OCF has three Kubernetes masters: (1) `deadlock`, (2) `coup`,
+and (3) `autocrat`. A Container Networking Interface (`cni`) is the last piece
+required for a working cluster. The `cni`'s purpose is to faciltate intra-pod
+communication. `puppetlabs-kubernetes` supports two choices: `weave` and
+`flannel`. Both solutions work out-the-box, and we've had success with
+`flannel` thus far so we've stuck with it.
+
+### Getting traffic into the cluster
+
+One of the challenges with running Kubernetes on bare-metal is getting traffic
+into the cluster. If you are running on `AWS`, `GCP`, or `Azure` you add `type:
+LoadBalancer` to your Service definitions and things just work. For bare-metal,
+this is an exercise left to the reader.
+
+The figure below demonstrates a request made for `templates.ocf.berkeley.edu`.
+
+```
+                                            -------------------------------------------------
+                                            |              Kubernetes Cluster               |
+                           HAProxy          |                                               |
+                          ----------        |                    Ingress         Pod        |
+                          |autocrat|        |                   ----------    ----------    |
+                          ----------        |                   |Worker1|     | ocfweb |    |
+                                            |                   ----------    ----------    |
+                                            |                                               |
+            *VIP*          HAProxy          |                    Ingress         Pod        |
+        ------------      ----------  ✘ SSL / Host: templates   ---------     -----------   |
+REQ --> |keepalived| -->  |deadlock|    -->   -------------->   |Worker1| --> |Templates|   |
```

Comment:
I'd try to make it more clear that the ingress that handles the request from haproxy doesn't necessarily run on the same worker with the templates pod.

Reply:
Ah, okay. I just using the Pod abstraction instead of a worker here, but yeah people new to Kube might not get the implication. I'll have to add it to the text though

Reply:
I think if you just add diagonal arrows in here to make it clear that the request could go to any of the pods from the ingress, then that would be fine. Same with the `keepalived` part, I think diagonal/sideways arrows would be good there to say it can go to any `HAProxy`/master, although I'm not sure if that would be confusing to suggest it's actually being split across masters or not. Maybe `keepalived` shouldn't be a separate step in the flow, but should just reside on one of the master servers? Like:

```
--------------------
|     deadlock:    |
|current keepalived|
|      master      |
--------------------
```

(or something similar)

Also `Worker1` -> `Worker2` on this line, `Worker1` is repeated twice, unless that was intentional because they could actually be scheduled onto the same worker.